cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java

```diff
@@ -0,0 +1,250 @@
+        kafkaCallbackSubscription.cancel();
+        if (!kafkaCallback.isRunning()) {
+          kafkaCallbacks.remove(topicPartition, kafkaCallback);
+        }
+      }
+    };
+  }
+
+  /**
+   * Callback class called when a Kafka message is received. The {@link #onReceived} method will
+   * extract the feed ID of the message received, and pass the notification encoded in the message
+   * to all handlers that are interested in that feed, using the {@code delegate} in-memory notification
+   * service.
+   * One callback is created per TopicPartition. It is created by the first subscription to a feed
+   * which maps to the TopicPartition.
+   */
+  private final class KafkaNotificationsCallback implements KafkaConsumer.MessageCallback, Cancellable {
+
+    private final TopicPartition topicPartition;
+    private final AtomicInteger subscriptions;
+
+    private Cancellable kafkaSubscription;
+    private boolean isRunning;
+
+    private KafkaNotificationsCallback(TopicPartition topicPartition) {
+      this.topicPartition = topicPartition;
+      this.subscriptions = new AtomicInteger(0);
+      this.isRunning = false;
+    }
+
+    public void start() {
+      KafkaConsumer.Preparer preparer = kafkaClient.getConsumer().prepare();
+
+      // TODO there is a bug in twill, that when the topic doesn't exist, add latest will not make subscription
+      // start from offset 0 - but that will be fixed soon
+      preparer.addLatest(topicPartition.getTopic(), topicPartition.getPartition());
+      kafkaSubscription = preparer.consume(this);
+
+      isRunning = true;
+    }
+
+    @Override
+    public void onReceived(Iterator<FetchedMessage> messages) {
+      int count = 0;
+      while (messages.hasNext()) {
+        count++;
+        FetchedMessage message = messages.next();
+        ByteBuffer payload = message.getPayload();
+
+        try {
+          KafkaMessage decodedMessage = KafkaMessageCodec.decode(payload);
+          try {
+            notificationReceived(KafkaNotificationUtils.getMessageFeed(decodedMessage.getMessageKey()),
+                                 decodedMessage.getNotificationJson());
+          } catch (Throwable t) {
+            LOG.warn("Error while processing notification {} with handler {}", decodedMessage.getNotificationJson(), t);
+          }
+        } catch (IOException e) {
+          LOG.error("Could not decode Kafka message {} using Gson.", message, e);
+        }
+      }
+      LOG.debug("Handled {} messages from kafka", count);
+    }
+
+    /**
+     * Add one more subscription to this {@link KafkaNotificationsCallback}.
+     *
+     * @return a {@link Cancellable} to cancel the subscription
+     */
+    public Cancellable addSubscription() {
+      subscriptions.incrementAndGet();
+      final KafkaNotificationsCallback that = this;
+      return new Cancellable() {
+        @Override
+        public void cancel() {
+          int i = subscriptions.decrementAndGet();
+          if (i == 0) {
```

Comment:
This logic is not race free. It's possible that the number of subscribers drop to 0, hence canceling the Kafka subscription, but this callback is still in the concurrent map while someone subscribe at the same time, causing the Kafka subscription gone but have an active notification subscriber.